### PR TITLE
 Supportability - expose more details about exceptions

### PR DESCRIPTION
**Pull request by bgavrilMS**

### CODE REVIEW

**Comment by henrik-me** [comment]
>         /// <param name="statusCode">Status code of the resposne received from the service.</param>

nit: response

---
Refers to: src/Microsoft.Identity.Client/MsalServiceException.cs:144 in 8e43cee. [](commit_id = 8e43cee787ad36a6c71f9d39d751fda7427bdca6, deletion_comment = False)

**Comment by henrik-me** [comment]
>         /// <param name="statusCode">Status code of the resposne received from the service.</param>

is this the http status code or some other status code? if not could we add HttpStatusCode as well here?

---
Refers to: src/Microsoft.Identity.Client/MsalServiceException.cs:144 in 8e43cee. [](commit_id = 8e43cee787ad36a6c71f9d39d751fda7427bdca6, deletion_comment = False)

File: src/Microsoft.Identity.Client/MsalServiceException.cs [hunk 1]
```diff
@@ -237,7 +237,14 @@ public MsalServiceException(string errorCode, string errorMessage, int statusCod
         /// <returns>A string representation of the current exception.</returns>
         public override string ToString()
         {
-            return base.ToString() + string.Format(CultureInfo.InvariantCulture, "\n\tStatusCode: {0}\n\tClaims: {1}", StatusCode, Claims);
+            return base.ToString() + string.Format(
```
Comment (henrik-me) [inline src/Microsoft.Identity.Client/MsalServiceException.cs]: >return [](start = 12, length = 6)

what about SubError, Error, CorrelationId and ? and errorcode ?

**Comment by henrik-me** [comment]
>                 sb.AppendLine(string.Format(CultureInfo.CurrentCulture, "Exception type: {0}", ex.GetType()));

what about adding the error message?

---
Refers to: src/Microsoft.Identity.Client/Exceptions/MsalExceptionFactory.cs:112 in 8e43cee. [](commit_id = 8e43cee787ad36a6c71f9d39d751fda7427bdca6, deletion_comment = False)

File: src/Microsoft.Identity.Client/Http/HttpManager.cs [hunk 2]
```diff
@@ -149,7 +149,7 @@ private HttpRequestMessage CreateRequestMessage(Uri endpoint, IDictionary<string
             }
             catch (TaskCanceledException exception)
             {
-                requestContext.Logger.ErrorPii(exception);
+                requestContext.Logger.Error("The HTTP request failed or it was cancelled. " + exception.Message);
```
Comment (henrik-me) [inline src/Microsoft.Identity.Client/Http/HttpManager.cs]: >cancelled [](start = 79, length = 9)

I believe American English this would be: canceled ?

**Comment by henrik-me** [comment]
>             string piiMessage = MsalExceptionFactory.GetPiiScrubbedExceptionDetails(msalException);

Nit: this should be the scrubbed message right? should this be piiScrubbedMessage? Not sure I understand why the message can't be part of the scrubbed message.

---
Refers to: tests/Microsoft.Identity.Test.Unit.net45/ExceptionTests/MsalExceptionFactoryTests.cs:155 in 8e43cee. [](commit_id = 8e43cee787ad36a6c71f9d39d751fda7427bdca6, deletion_comment = False)

File: tests/Microsoft.Identity.Test.Unit.net45/ExceptionTests/MsalExceptionTests.cs [hunk 3]
```diff
@@ -47,12 +49,59 @@ public void ExceptionsArePubliclyCreatable_MsalException()
...
+            };
+
+            // Act
+            var ex = MsalExceptionFactory.GetServiceException(
+                "errCode",
+                "errMessage",
+                httpResponse,
+                innerException);
+
+            Assert.IsTrue(ex.ToString().Contains("errCode"));
```
Comment (henrik-me) [inline tests/Microsoft.Identity.Test.Unit.net45/ExceptionTests/MsalExceptionTests.cs]: nit: given you have Arrange and Act... you should perhaps add //Assert here?

**Comment by henrik-me** [comment]
ok. a few comments which we can always follow up on, but I believe it would be good to re-assess a few things (goal: make developers even better able to self diagnose the root cause of the issue). Perhaps we can also add a link to aka.ms/how-to-trouble-shoot-msal-exceptions ... a page which may or may not yet exist (JM did add something previously) where we can continue to provide guidance on various error scenarios and how to handle those.

File: src/Microsoft.Identity.Client/Http/HttpManager.cs (hunk 2 above)
Comment (bgavrilMS) [reply src/Microsoft.Identity.Client/Http/HttpManager.cs]: Oh, you are breaking my bones here Henrik. Ok ...

File: src/Microsoft.Identity.Client/MsalServiceException.cs (hunk 1 above)
Comment (bgavrilMS) [reply src/Microsoft.Identity.Client/MsalServiceException.cs]: They are all in the ResponseBody as a json, it's pretty readable as is. Actually we also don't need claims.

Comment (bgavrilMS) [reply src/Microsoft.Identity.Client/MsalServiceException.cs]: base.ToString() will display error code and error message. SubError, CorrelationID etc. are contained in the ResponseBody, and double logging creates confusion. The unit test does assert all of these are present.

**Comment by bgavrilMS** [comment]
Addressed all but one issue - include the exception message in the non-pii logs. I can think of URLs leaking out this way, not sure about email addresses / UPNs, but it is possible. @jmprieur - thoughts?

**Pull request merged**